subsys/net/lib/sockets/sockets.c

```diff
@@ -870,7 +917,16 @@ static inline ssize_t zsock_recv_dgram(struct net_context *ctx,
 	if (flags & ZSOCK_MSG_PEEK) {
 		int res;
 
+		if (K_TIMEOUT_EQ(timeout, K_FOREVER)) {
+			sys_mutex_unlock(lock);
+		}
+
 		res = k_fifo_wait_non_empty(&ctx->recv_q, timeout);
+
+		if (K_TIMEOUT_EQ(timeout, K_FOREVER)) {
+			(void)sys_mutex_lock(lock, K_FOREVER);
```

Comment:
I think there might be a problem here... The socket might've gotten closed when `recv` was waiting on a FIFO, so technically we'll use a mutex of an unassigned FD object here. Or even worse, if the FD was reassigned, we'll use someone else mutex.

I'm not really sure how to solve that. Theoretically, we should re-verify that we still own the FD, but this might be tricky given we have no knowledge of the FD we use at this point.

Reply:
The refcounting PR at #25902 might help here.